### PR TITLE
Features: Data Governance

### PR DESCRIPTION
**Pull request by igorescobar**
Addresses #13 

1. Adding mix format CI validation.
1. Validates the inner config options `"p"."u"` and `"p"."r"`.
1. If the user decides to set its own base_url, data governance is ignored unless `"p"."r"` is FORCE.
1. Converted a few inner data structures to Map instead of lists so we can handle user overwrites more easily.
1. I tried to do all Map manipulations with O(1) time complexity.
1. Not using anymore the address `https://cdn.configcat.com` in favor of `cdn-eu` and `cdn-global`.
1. Test cases matches the Ruby SDK

### CODE REVIEW

File: lib/config_cat/data_governance.ex [hunk 1]
```diff
@@ -0,0 +1,11 @@
+defmodule ConfigCat.DataGovernance do
```
Comment (laliconfigcat) [inline lib/config_cat/data_governance.ex:1]: Hmm, it's not that easy to define an enum.. :D

File: test/config_cat/config_fetcher_test.exs [hunk 2]
```diff
@@ -164,6 +190,110 @@ defmodule ConfigCat.ConfigFetcherTest do
     {:ok, _} = ConfigFetcher.fetch(fetcher)
   end
 
+  describe "datagovernance behaviour" do
```
Comment (laliconfigcat) [inline test/config_cat/config_fetcher_test.exs]: Could you please extract the datagovernance tests to a data_governance_test.exs? It'll be easier for us to find the related code later if we use the same terminology as the other SDKs. 
Sorry about this, but maintaining 10+ different languages is really hard and it makes our life much easier if we can find things where we expect them.

File: lib/config_cat/config_fetcher.ex [hunk 3]
```diff
@@ -107,7 +154,34 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
        when code >= 200 and code < 300 do
-    with etag <- extract_etag(headers) do
+    with etag <- extract_etag(headers),
+         %{base_url: new_base_url, custom_endpoint: custom_endpoint} <-
+           Map.take(state, [:base_url, :custom_endpoint]),
+         p <- Map.get(config, Constants.preferences(), %{}),
+         base_url <- Map.get(p, Constants.preferences_base_url()),
+         redirect <- Map.get(p, Constants.redirect()) do
+      state =
+        cond do
```
Comment (laliconfigcat) [inline lib/config_cat/config_fetcher.ex:148]: There is a missing part of this implementation. We prepared our SDKs to handle cases when we accidentally mess up our config.jsons. So to prevent redirect loops we only retry the fetch 3 times and if there is a redirect loop, we return with the last fetch's response and log a warning. 
https://github.com/configcat/ruby-sdk/blob/master/lib/configcat/configfetcher.rb#L120

Example:
cdn-global.configcat.com/.../config_v5.json:    { p: { r: 1, u: 'cdn-eu.configcat.com' }}
cdn-eu.configcat.com/.../config_v5.json:   { p: { r: 1, u: 'cdn-global.configcat.com' }}

File: test/config_cat/config_fetcher_test.exs [hunk 4]
```diff
@@ -164,6 +190,110 @@ defmodule ConfigCat.ConfigFetcherTest do
...
+        {:ok, %Response{status_code: 200, body: config}}
+      end)
+      |> expect(:get, 1, fn ^redirect_url, _headers, [] ->
+        {:ok, %Response{status_code: 200, body: config}}
+      end)
+
+      assert {:ok, ^config} = ConfigFetcher.fetch(fetcher)
+    end
+  end
+
```
Comment (laliconfigcat) [inline test/config_cat/config_fetcher_test.exs]: I miss some cases here.
This is a really important and crucial part of the SDKs. Please implement all of the tests from https://github.com/configcat/ruby-sdk/blob/master/spec/datagovernance_spec.rb

**Review by laliconfigcat — CHANGES_REQUESTED** [review]

File: lib/config_cat/data_governance.ex (hunk 1 above)
Comment (igorescobar) [reply lib/config_cat/data_governance.ex:1]: Typespecs makes it a bit bloated 😄

File: test/config_cat/config_fetcher_test.exs (hunk 2 above)
Comment (igorescobar) [reply test/config_cat/config_fetcher_test.exs]: @randycoulman Do you have a better solution for this? 
We usually have One file -> One spec. For now, I will just go ahead and create the file but to me sounds wrong 😅 if you have a better suggestion so everybody could be happy I'm all ears.

File: lib/config_cat/config_fetcher.ex (hunk 3 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex:148]: With the current implementation, I don't see this ever happening. We only follow the link if it's different then the link followed previously. And this is currently being validated with the current unit tests when we verify the number of calls even when the response is exactly the same as the one prior.

Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex:148]: Ah! Now I see what you mean... The config files itself are not configured properly so they will be redirecting to each other!

**Comment by randycoulman** [comment]
@igorescobar I don't have a problem with a separate spec file for data governance.  There seems to be a general convention of one spec file per implementation file in the TDD world, but I think that's more an accident of tooling support rather than a conscious design choice.

File: lib/config_cat.ex [hunk 5]
```diff
@@ -21,6 +22,7 @@ defmodule ConfigCat do
           | {:cache, module()}
           | {:cache_policy, CachePolicy.t()}
           | {:http_proxy, String.t()}
+          | {:data_governance, DataGovernance.t()}
```
Comment (randycoulman) [inline lib/config_cat.ex:24]: Very minor: I've been trying to keep things alphabetically-sorted to make them easier to find.  This shows up here and a few other places below. Not a big deal, though.

File: lib/config_cat/config_fetcher.ex [hunk 6]
```diff
@@ -7,15 +7,23 @@ defmodule ConfigCat.ConfigFetcher do
   @type result :: {:ok, Config.t()} | {:ok, :unchanged} | fetch_error()
 
   @callback fetch(id()) :: result()
+
+  defmodule RedirectMode do
+    defmacro no_redirect, do: 0
+    defmacro should_redirect, do: 1
+    defmacro force_redirect, do: 2
+  end
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex:15]: I'd like to find a better way to handle constants throughout the app, but that can be a future thing.

Again, though, this can be a follow-up refactoring to clean up constants a bit.

File: lib/config_cat/config_fetcher.ex [hunk 7]
```diff
@@ -34,14 +43,49 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
+        %{state | custom_endpoint: true}
+
+      _ ->
+        state
+    end
+  end
+
+  defp data_governance_options(state, options) do
+    with %{data_governance: data_governance, custom_endpoint: custom_endpoint} <-
+           Map.take(state, [:data_governance, :custom_endpoint]) do
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: #minor No need for the `Map.take` here; you can pattern match directly against `state` to get the values you need.  

```elixir
with %{data_governance: data_governance, custom_endpoint: custom_endpoint} <- state do
```

File: lib/config_cat/config_fetcher.ex [hunk 8]
```diff
@@ -55,6 +99,10 @@ defmodule ConfigCat.CacheControlConfigFetcher do
 
   @impl GenServer
   def handle_call(:fetch, _from, state) do
+    handle_fetch_call(state)
+  end
+
+  defp handle_fetch_call(state) do
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: #minor: Can we name this with something other than a `handle_` prefix? I find that it mixes in too well with the real `handle_call`/`handle_fetch` functions.

I used `do_*` naming in another part of the codebase, so maybe `do_fetch` here?

File: lib/config_cat/config_fetcher.ex [hunk 9]
```diff
@@ -107,7 +155,43 @@ defmodule ConfigCat.CacheControlConfigFetcher do
 
   defp handle_response(%Response{status_code: code, body: config, headers: headers}, state)
        when code >= 200 and code < 300 do
-    with etag <- extract_etag(headers) do
+    with etag <- extract_etag(headers),
+         %{base_url: new_base_url, custom_endpoint: custom_endpoint, redirects: redirects} <-
+           Map.take(state, [:base_url, :custom_endpoint, :redirects]),
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: Again, no need for a `Map.take` call here.

File: lib/config_cat/config_fetcher.ex [hunk 10]
```diff
@@ -107,7 +155,43 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
+         p <- Map.get(config, Constants.preferences(), %{}),
+         base_url <- Map.get(p, Constants.preferences_base_url()),
+         redirect <- Map.get(p, Constants.redirect()) do
+      state =
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: Should this be named `new_state` to avoid confusion with the incoming state?

File: lib/config_cat/config_fetcher.ex [hunk 11]
```diff
@@ -107,7 +155,43 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
+        cond do
+          custom_endpoint && redirect != RedirectMode.force_redirect() ->
+            state
+
+          redirect == RedirectMode.no_redirect() ->
+            state
+
+          base_url && !Map.has_key?(redirects, new_base_url) ->
+            {_, {_, _}, state} =
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: #minor: Could be simplified a bit to `{_, _, state}`

File: lib/config_cat/data_governance.ex [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  @type t :: global | eu_only
+
+  @type eu_only :: integer()
+  @type global :: integer()
```
Comment (randycoulman) [inline lib/config_cat/data_governance.ex]: This doesn't match the implementation below. I haven't done constants this way before, but maybe we should do something like `@type t :: 0 | 1` or maybe:

```elixir
@type eu_only :: 1
@type global :: 0
@type t :: global | eu_only

@spec global :: global()
defmacro global, do: 0

@spec eu_only :: eu_only()
defmacro eu_only, do: 1
```

?

File: test/config_cat/config_fetcher_test.exs [hunk 13]
```diff
@@ -7,7 +7,7 @@ defmodule ConfigCat.ConfigFetcherTest do
   alias ConfigCat.{Constants, MockAPI}
   alias HTTPoison.Response
 
-  require ConfigCat.Constants
+  require ConfigCat.{Constants}
```
Comment (randycoulman) [inline test/config_cat/config_fetcher_test.exs:10]: #minor No need for `{` ... `}` here

File: test/config_cat/config_fetcher_test.exs [hunk 14]
```diff
@@ -38,8 +44,7 @@ defmodule ConfigCat.ConfigFetcherTest do
   test "successful fetch", %{config: config, sdk_key: sdk_key} = context do
     {:ok, fetcher} = start_fetcher(context)
 
-    url =
-      "#{Constants.base_url()}/#{Constants.base_path()}/#{sdk_key}/#{Constants.config_filename()}"
+    url = global_config_url(sdk_key)
```
Comment (randycoulman) [inline test/config_cat/config_fetcher_test.exs]: Nice! 👍

File: test/config_cat/config_fetcher_test.exs [hunk 15]
```diff
@@ -137,10 +142,10 @@ defmodule ConfigCat.ConfigFetcherTest do
   end
 
   test "allows base URL to be configured", %{config: config, sdk_key: sdk_key} = context do
-    base_url = "https://BASE_URL/"
+    base_url = "https://BASE_URL"
```
Comment (randycoulman) [inline test/config_cat/config_fetcher_test.exs]: This was really subtle, but was testing that the URL construction code in ConfigFetcher worked properly with a trailing `/` on it, because there's a good chance that callers will include one.

If you want, we could sanitize this early on instead, but I do want to make sure we handle this case.

File: test/config_cat/data_governance_test.exs [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+  setup do
+    {:ok,
+     %{
+       redirect_base_url: "https://redirect.configcat.com",
+       forced_base_url: "https://forced.configcat.com",
+       custom_base_url: "https://custom.configcat.com",
+       etag: "ETAG",
+       mode: "m",
+       sdk_key: "SDK_KEY"
+     }}
```
Comment (randycoulman) [inline test/config_cat/data_governance_test.exs]: #optional We could define all of these as module attributes (`@mode "m"`) so that we don't have to list them in every test that uses them.  We could make the same change in the ConfigFetcher tests as well.  No need to do this as part of this PR, but I might do it in a cleanup PR later since I made the original mess that you're following here. 🤦

File: test/config_cat/data_governance_test.exs [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+  end
+
+  defp start_fetcher(%{mode: mode, sdk_key: sdk_key}, options) do
+    name = UUID.uuid4() |> String.to_atom()
+    default_options = [api: MockAPI, mode: mode, name: name, sdk_key: sdk_key]
+    options = Keyword.merge(default_options, options)
```
Comment (randycoulman) [inline test/config_cat/data_governance_test.exs]: We're doing the merge here and again below on line 35. I think the one below is enough. And... it turns out I did the same thing in the ConfigFetcher tests 🤦 .

File: test/config_cat/data_governance_test.exs [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+      {:ok, %Response{status_code: 200, body: config_to_global}}
+    end)
+
+    assert {:ok, _} = ConfigFetcher.fetch(fetcher)
+    assert {:ok, _} = ConfigFetcher.fetch(fetcher)
+  end
+
+  defp stub_response(response_uri, redirect) do
+    %{
+      "f" => %{"test" => "json"},
```
Comment (randycoulman) [inline test/config_cat/data_governance_test.exs]: Should this use `Constants.feature_flags` for consistency with the following few lines?

File: test/integration_test.exs [hunk 19]
```diff
@@ -1,7 +1,8 @@
 defmodule ConfigCat.IntegrationTest do
   use ExUnit.Case, async: true
 
-  alias ConfigCat.CachePolicy
+  require ConfigCat.{DataGovernance}
```
Comment (randycoulman) [inline test/integration_test.exs]: #minor No `{`...`}` needed here.

File: lib/config_cat/config_fetcher.ex [hunk 20]
```diff
@@ -34,14 +43,49 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
+          %{state | base_url: base_url}
+
+        data_governance == DataGovernance.eu_only() ->
+          %{state | base_url: Constants.base_url_eu_only()}
+
+        true ->
+          state
+      end
+    end
+  end
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: I'm finding this option processing code quite hard to follow. Here are a few things that I'm having trouble with:

- It feels like the responsibilities of the various helpers could be separated a bit
- The mix of maps and keyword lists is a bit confusing
- `redirects` is in `default_options`, but isn't an option at all; it's only part of the GenServer state
- Having an argument named `state` in code that's processing options is confusing in a GenServer, which also has a `state`.

I'd like to propose the following refactoring, but it's up to you whether or not you want to take the suggestion. 

Here's what I did:
- Remove `base_url` from `default_options` entirely so that we can merge in the default options early, but still determine whether or not a `base_url` was provided by the caller.
- Process the options as a keyword list until we convert to a Map to become the initial_state
- Start with the same option processing code all of the other GenServers use (`default_options() |> Keyword.merge(options)`)
- Add a default base_url when necessary, and set the `custom_endpoint` flag at the same time
- Add additional state (`etag` and `redirects`) at the end

```elixir
  @spec start_link(options()) :: GenServer.on_start()
  def start_link(options) do
    {name, options} = Keyword.pop!(options, :name)

    initial_state =
      default_options()
      |> Keyword.merge(options)
      |> choose_base_url()
      |> Map.new()
      |> Map.merge(%{etag: nil, redirects: %{}})

    GenServer.start_link(__MODULE__, initial_state, name: name)
  end

  defp default_options,
    do: [api: ConfigCat.API, data_governance: DataGovernance.global()]

  defp choose_base_url(options) do
    case Keyword.get(options, :base_url) do
      nil ->
        base_url = options |> Keyword.get(:data_governance) |> default_url()
        Keyword.merge(options, base_url: base_url, custom_endpoint: false)

      _ ->
        Keyword.put(options, :custom_endpoint, true)
    end
  end

  defp default_url(DataGovernance.eu_only()), do: Constants.base_url_eu_only()
  defp default_url(_), do: Constants.base_url_global()
```

File: lib/config_cat/config_fetcher.ex [hunk 21]
```diff
@@ -34,14 +43,49 @@ defmodule ConfigCat.CacheControlConfigFetcher do
...
     GenServer.start_link(__MODULE__, initial_state, name: name)
   end
 
-  defp default_options, do: [api: ConfigCat.API, base_url: Constants.base_url()]
+  defp default_options,
+    do: %{
+      api: ConfigCat.API,
+      base_url: Constants.base_url_global(),
+      data_governance: DataGovernance.global(),
+      custom_endpoint: false,
```
Comment (randycoulman) [inline lib/config_cat/config_fetcher.ex]: Should we name this `custom_endpoint?` to make it clearer that this is a boolean?

**Review by randycoulman — APPROVED** [review]
Wow, this is such a confusing feature! I had to read the Ruby code a couple of times to understand what was going on, and I'm still not sure I've got it fully nailed down.  You've done a really nice job capturing all of that here!

I've left a lot of feedback below, but all of it is optional.

The one thing we should double-check with @laliconfigcat is whether we need to start the redirects over again on subsequent calls to `fetch`; if we do, then this implementation will have to change accordingly. But if not, then you can implement whichever of my suggestions you want and leave the rest.

I'm hoping my review doesn't sound overly critical; there were some parts of the code I had to really work at to understand, so I've made some suggestions in the hopes that it's easier for the next new reader who comes along.  I'm fully aware that I'll understand code differently than others which is why all of my suggestions are optional.

File: lib/config_cat/config_fetcher.ex (hunk 6 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex:15]: I added it here because it's strictly related to the config fetcher and this is kind of what they are doing on their Ruby SDK too.

File: lib/config_cat/config_fetcher.ex (hunk 7 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: For some reason, it wasn't working when I tried... but now I tried again and it worker 😅

File: lib/config_cat/config_fetcher.ex (hunk 8 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: sure!

File: lib/config_cat/config_fetcher.ex (hunk 9 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: Done!

File: test/config_cat/data_governance_test.exs (hunk 17 above)
Comment (igorescobar) [reply test/config_cat/data_governance_test.exs]: haha! I didn't notice! Fixed!

File: test/config_cat/data_governance_test.exs (hunk 18 above)
Comment (igorescobar) [reply test/config_cat/data_governance_test.exs]: Removed it, it wasn't necessary for the scope of this test.

File: test/integration_test.exs (hunk 19 above)
Comment (igorescobar) [reply test/integration_test.exs]: Done

File: lib/config_cat/config_fetcher.ex (hunk 20 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: Looks great! :)

File: lib/config_cat/config_fetcher.ex (hunk 21 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: Sure!

File: lib/config_cat.ex (hunk 5 above)
Comment (igorescobar) [reply lib/config_cat.ex:24]: done!

File: lib/config_cat/config_fetcher.ex (hunk 10 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: done!

File: lib/config_cat/config_fetcher.ex (hunk 11 above)
Comment (igorescobar) [reply lib/config_cat/config_fetcher.ex]: done!

File: lib/config_cat/data_governance.ex (hunk 12 above)
Comment (igorescobar) [reply lib/config_cat/data_governance.ex]: Done 👍

File: test/config_cat/config_fetcher_test.exs (hunk 13 above)
Comment (igorescobar) [reply test/config_cat/config_fetcher_test.exs:10]: Done 👍

**Review by laliconfigcat — APPROVED** [review]

File: test/config_cat/config_fetcher_test.exs (hunk 15 above)
Comment (igorescobar) [reply test/config_cat/config_fetcher_test.exs]: I completely missed that. I added a comment on top of it for the future-me and fixed the test.

File: test/config_cat/data_governance_test.exs (hunk 16 above)
Comment (igorescobar) [reply test/config_cat/data_governance_test.exs]: Done! On both files.
